cli/azd/cmd/get_access_token.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/azure/azure-dev/cli/azd/cmd/actions"
+	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/contracts"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+type getAccessTokenFlags struct {
+	outputFormat string
+	scopes       []string
+	global       *internal.GlobalCommandOptions
+}
+
+func getAccessTokenCmdDesign(global *internal.GlobalCommandOptions) (*cobra.Command, *getAccessTokenFlags) {
+	cmd := &cobra.Command{
+		Use:    "get-access-token",
+		Short:  "Get an access token for the current signed in principal.",
+		Hidden: true,
+	}
+	getAccessTokenFlags := &getAccessTokenFlags{}
+	getAccessTokenFlags.Bind(cmd.Flags(), global)
+	return cmd, getAccessTokenFlags
+}
+
+func (f *getAccessTokenFlags) Bind(local *pflag.FlagSet, global *internal.GlobalCommandOptions) {
+	f.global = global
+	output.AddOutputFlag(local, &f.outputFormat, []output.Format{output.JsonFormat}, output.NoneFormat)
+	local.StringArrayVar(&f.scopes, "scope", nil, "The scope to use when requesting an access token")
+}
+
+type getAccessTokenAction struct {
+	credential azcore.TokenCredential
+	formatter  output.Formatter
+	writer     io.Writer
+	flags      getAccessTokenFlags
+}
+
+func newGetAccessTokenAction(
+	credential azcore.TokenCredential,
+	formatter output.Formatter,
+	writer io.Writer,
+	flags getAccessTokenFlags,
+) *getAccessTokenAction {
+	return &getAccessTokenAction{
+		credential: credential,
+		formatter:  formatter,
+		writer:     writer,
+		flags:      flags,
+	}
+}
+
+func (a *getAccessTokenAction) Run(ctx context.Context) (*actions.ActionResult, error) {
+	token, err := a.credential.GetToken(ctx, policy.TokenRequestOptions{
+		Scopes: a.flags.scopes,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("fetching token: %w", err)
+	}
+
+	res := contracts.GetAccessTokenResult{
+		Token:     token.Token,
+		ExpiresOn: token.ExpiresOn,
+	}
+
+	return nil, a.formatter.Format(res, a.writer, nil)
```

Comment:
why are we using the error for the result?  is that the pattern for hidden commands?

Reply:
Not sure I understand the question, but it may be that I don't have an understanding of the new model for Actions.

Previously, our actions just returned an error, so code that would use the formatter to write results would do something like this:

```golang
return a.formatter.Format(res, a.writer, nil)
```

When stuff got refactored to introduce the new signatures, a bunch of existing code that followed this pattern got updated to return a `nil` result and the error, as before.

I don't think there's a sensible `ActionResult` object I can return here (looking at what's in my tree it seems like this type is still highly specialized to for the interactive UI output scenario, not returning JSON data) so I return a nil one, and any error that I got from doing the format of my result.

Does that make sense?

FWIW  - we do this sort of thing a few other places (`login.go`, `show.go`, `templates.go`).

If there's some other pattern to follow, happy to do it.

Reply:
haaa,, never mind!!  I got it now!  the `a.formatter.Format()` takes care of `formatting` AND `printing to console. And it returns `err`.
This is somehow confusing to me (now I remember that I saw this in the azd env commands).  I would usually expect a function named `Format()` to return either a string or some object representing the formatted input.  

So, I was getting a little lost trying to find the code to do the command output.  
Maybe we should rename `Format()` to `FormatAndWrite()`. Right now, the only hint about it writes the output instead of returning it, is by looking at the input arg `writter`.  But that's totally orthogonal to this PR.  

Thank you for explaining.  Resolving this comment now.